util/device_sw_utils/extract_sw_logs.py

```diff
@@ -0,0 +1,217 @@
+    doing so if a neline is encountered in the middle of a string, it ends up
+    adding that newline in the output files this script generates. The output
+    of this script is consumed by a monitor written in SystemVerilog
+    (hw/dv/sv/sw_logger_if), a language with limited parsing / processing
+    support.'''
+    return re.sub(r"\r+", "\r", string.replace("\n", "\r")).strip()
+
+
+def cleanup_format(_format):
+    '''This function converts C style format specifiers to SV style.
+
+    It makes the folllowing substitutions:
+    - Change %[N]?i, %[N]?u --> %[N]?d
+    - Change %[N]?x, %[N]?p --> %[N]?h
+    - Change %[N]?X --> %[N]?H
+
+    The below is a non-standard format specifier added in OpenTitan
+    (see sw/device/lib/base/print.c for more details). A single %z specifier
+    consumes 2 arguments instead of 1 and hence has to converted as such to
+    prevent the log monitor in SystemVerilog from throwing an error at runtime.
+    - Change %[N]?z --> %[N]?s[%d].'''
+    # TODO: FIXME
+    return cleanup_newlines(_format)
+
+
+def prune_filename(filename):
+    'This function prunes the filename to only display the hierarchy under sw/'
+    hier = "sw/device"
+    index = filename.find(hier)
+    return (filename if index == -1 else filename[index:])
+
+
+def get_addr_strings(ro_contents):
+    '''This function processes the read-only sections of the elf supplied as
+    a list of ro_content tuples comprising of base addr, size and data in bytes
+    and converts it into an {addr: string} dict which is returned. Non ASCII
+    characters if encountered are replaced with ?.'''
+    result = {}
+    for ro_content in ro_contents:
+        str_start = 0
+        base_addr, size, data = ro_content
+        while (str_start < size):
+            str_end = data.find(b'\0', str_start)
+            # Skip if start and end is the same
+            if str_start == str_end:
+                str_start += 1
+                continue
+            if str_end == -1: str_end = size - 1
+            # Get full string address by adding base addr to the start.
+            addr = base_addr + str_start
+            string = data[str_start:str_end].decode('utf-8', errors='replace')
+            string = cleanup_newlines(string)
+            if addr in result:
+                print("Error: duplicate {addr: string} pair encountered:\n")
+                print("  Old: addr: {}\nstring: {}".format(addr, result[addr]))
+                print("  New: addr: {}\nstring: {}".format(addr, string))
+                # This is as such, not a fatal error - sim can still proceed
+                # without any issues. However, it can make debug difficult
+                # since incorrect logs will get printed. It is better to fix
+                # it proactively.
+                sys.exit(1)
+            result[addr] = string
+            str_start = str_end + 1
+    return result
+
+
+def get_str_at_addr(str_addr, addr_strings):
```

Comment:
The strings can be looked up without splitting the data into individual parts first by using the address only. This code is significantly simpler than what you have currently -- as long as you don't need the split strings. Since they are only used in the rodata.txt file, the question remains: what's the purpose of that file? I don't see it used anywhere.

Reply:
The flow currently does not support `%s` format specifier yet. Dumping the complete rodata is a step towards enabling it. It will be done in a subsequent PR since this is already large.